NPE during compilation

I get an error during compilation

```
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: .../MyModel.kt
The root cause java.lang.NullPointerException was thrown at: net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
```

I'm confused because I'm not using the `@Secret` annotation on this class - maybe it's because of the inline value class, `ModelId`?


```kotlin
data class MyModel(
    override val id: ModelId,
    override val name: String,
    override val description: String?,
) : Product


interface Product {
    val id: ModelId?
    val name: String?
    val description: String?
}


@JvmInline
value class ModelId(val value: String) : CharSequence by value
```

* Kotlin 1.7.0
* Sekret 0.1.2



```none
> Task :my-application:compileKotlin FAILED
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: .../MyModel.kt
The root cause java.lang.NullPointerException was thrown at: net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:239)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:235)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:68)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:55)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:41)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:29)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:43)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.invokeCodegen(JvmIrCodegenFactory.kt:284)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.runCodegen(KotlinToJVMBytecodeCompiler.kt:355)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:136)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:60)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:157)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:94)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:43)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:477)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:127)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally(IncrementalCompilerRunner.kt:366)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally$default(IncrementalCompilerRunner.kt:311)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl$rebuild(IncrementalCompilerRunner.kt:111)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:165)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:74)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:625)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:101)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1739)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
	at net.afanasev.sekret.kotlin.SekretClassBuilder.done(SekretClassBuilder.kt:89)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generate(ClassCodegen.kt:195)
	at org.jetbrains.kotlin.backend.jvm.FileCodegen.lower(JvmPhases.kt:44)
	at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:120)
	at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:116)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:65)
	... 42 more
```

--- a/.editorconfig
+++ b/.editorconfig
@@ -5,3 +5,4 @@ indent_size = 4
 indent_style = space
 insert_final_newline = true
 end_of_line = lf
+max_line_length = 200

--- a/buildSrc/build.gradle.kts
+++ b/buildSrc/build.gradle.kts
@@ -25,10 +25,6 @@ tasks.withType<KotlinCompile>().configureEach {
 }
 
 kotlin {
-    jvmToolchain {
-        (this as JavaToolchainSpec).languageVersion.set(JavaLanguageVersion.of(8))
-    }
-
     kotlinDslPluginOptions {
         jvmTarget.set("1.8")
     }

--- a/buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts
+++ b/buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts
@@ -7,7 +7,7 @@ plugins {
 
 kotlin {
     jvmToolchain {
-        (this as JavaToolchainSpec).languageVersion.set(JavaLanguageVersion.of("8"))
+        languageVersion.set(JavaLanguageVersion.of("8"))
     }
 }
 

--- a/buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts
+++ b/buildSrc/src/main/kotlin/buildsrc/convention/kotlin-jvm.gradle.kts
@@ -6,9 +6,6 @@ plugins {
 }
 
 kotlin {
-    jvmToolchain {
-        languageVersion.set(JavaLanguageVersion.of("8"))
-    }
 }
 
 java {

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/FieldInfo.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/FieldInfo.kt
@@ -1,0 +1,7 @@
+package net.afanasev.sekret.kotlin
+
+data class FieldInfo(
+    val desc: String,
+    val needToHide: Boolean,
+    val isNullable: Boolean,
+)

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/FieldInfo.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/FieldInfo.kt
@@ -4,4 +4,5 @@ data class FieldInfo(
     val desc: String,
     val needToHide: Boolean,
     val isNullable: Boolean,
+    val wrapperClassName: String?,
 )

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
@@ -15,8 +15,6 @@ import org.jetbrains.org.objectweb.asm.Label
 import org.jetbrains.org.objectweb.asm.MethodVisitor
 import org.jetbrains.org.objectweb.asm.Opcodes
 
-typealias FieldInfo = Triple<String, Boolean, Boolean>
-
 class SekretClassBuilder(
     internal val classBuilder: ClassBuilder,
     annotations: List<String>,
@@ -61,8 +59,14 @@ class SekretClassBuilder(
             generateToString = true
             // skipping toString generation as it will be constructed manually in done() function.
             object : MethodVisitor(Opcodes.ASM5) {
-                override fun visitFieldInsn(opcode: Int, owner: String, name: String, descriptor: String) {
+                override fun visitFieldInsn(
+                    opcode: Int,
+                    owner: String,
+                    name: String,
+                    descriptor: String
+                ) {
                     if (opcode == Opcodes.GETFIELD) {
+                        println("skipping $owner/$name/$descriptor - opcode $opcode=='GET FIELD'")
                         fields[name] = null
                     }
                 }
@@ -74,8 +78,13 @@ class SekretClassBuilder(
                     descriptor: String,
                     isInterface: Boolean
                 ) {
-                    if (opcode == Opcodes.INVOKEVIRTUAL && owner == origin.classDescriptor() && name.startsWith("get")) {
-                        fields[name.substring(3).replaceFirstChar { it.lowercase() }] = null
+                    if (opcode == Opcodes.INVOKEVIRTUAL
+                        && owner == origin.classDescriptor()
+                        && name.startsWith("get")
+                    ) {
+                        val keyName = name.substring(3).replaceFirstChar { it.lowercase() }
+                        println("skipping $owner/$keyName/$descriptor - opcode $opcode=='INVOKE VIRTUAL'. interface:$isInterface")
+                        fields[keyName] = null
                     }
                 }
             }
@@ -102,23 +111,23 @@ class SekretClassBuilder(
         mv.visitLdcInsn("${thisName.split("/").last()}(")
         appendToStringBuilder(mv)
 
-        fields.onEachIndexed { index, entry ->
-            val name = entry.key
-            val fieldInfo = entry.value!!
-            val desc = fieldInfo.first
-            val needToHide = fieldInfo.second
-            val isNullable = fieldInfo.third
+        fields.onEachIndexed { index, (name, fieldInfo) ->
+
+            if (fieldInfo == null) {
+                println("WARNING $index/$name has no field info")
+                return@onEachIndexed
+            }
 
             mv.visitLdcInsn((if (index == 0) "" else ", ") + "$name=")
             appendToStringBuilder(mv)
 
-            if (needToHide) {
-                if (maskNulls && isNullable) {
+            if (fieldInfo.needToHide) {
+                if (maskNulls && fieldInfo.isNullable) {
                     val loadNullLabel = Label()
                     val appendStringLabel = Label()
 
                     mv.visitVarInsn(Opcodes.ALOAD, 0)
-                    mv.visitFieldInsn(Opcodes.GETFIELD, thisName, name, desc)
+                    mv.visitFieldInsn(Opcodes.GETFIELD, thisName, name, fieldInfo.desc)
                     mv.visitJumpInsn(Opcodes.IFNULL, loadNullLabel)
 
                     mv.visitLdcInsn(mask)
@@ -135,27 +144,28 @@ class SekretClassBuilder(
                 }
             } else {
                 mv.visitVarInsn(Opcodes.ALOAD, 0)
-                mv.visitFieldInsn(Opcodes.GETFIELD, thisName, name, desc)
+                mv.visitFieldInsn(Opcodes.GETFIELD, thisName, name, fieldInfo.desc)
 
                 when {
                     // primitives or descriptors supported by StringBuilder
-                    desc.length == 1 || STRING_BUILDER_SUPPORTED_DESCRIPTORS.contains(desc) -> {
-                        appendToStringBuilder(mv, desc)
+                    fieldInfo.desc.length == 1
+                        || STRING_BUILDER_SUPPORTED_DESCRIPTORS.contains(fieldInfo.desc) -> {
+                        appendToStringBuilder(mv, fieldInfo.desc)
                     }
                     // arrays
-                    desc[0] == '[' -> {
-                        val isPrimitiveArray = desc.length == 2
+                    fieldInfo.desc[0] == '['                                             -> {
+                        val isPrimitiveArray = fieldInfo.desc.length == 2
                         mv.visitMethodInsn(
                             Opcodes.INVOKESTATIC,
                             "java/util/Arrays",
                             "toString",
-                            "(${if (isPrimitiveArray) desc else "[Ljava/lang/Object;"})Ljava/lang/String;",
+                            "(${if (isPrimitiveArray) fieldInfo.desc else "[Ljava/lang/Object;"})Ljava/lang/String;",
                             false,
                         )
                         appendToStringBuilder(mv)
                     }
                     // others go as Object
-                    else -> {
+                    else                                                                 -> {
                         appendToStringBuilder(mv, "Ljava/lang/Object;")
                     }
                 }

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
@@ -26,6 +26,8 @@ class SekretClassBuilder(
     private val fields = linkedMapOf<String, FieldInfo?>()
     private var generateToString = false
 
+    private var error = ""
+
     override fun getDelegate(): ClassBuilder = classBuilder
 
     override fun newField(
@@ -40,7 +42,11 @@ class SekretClassBuilder(
             (origin.descriptor as? PropertyDescriptor)?.let { descriptor ->
                 val hide = annotations.any { descriptor.annotations.hasAnnotation(it) }
                 if (fields.contains(name)) {
-                    fields[name] = FieldInfo(desc, hide, descriptor.type.isNullable())
+                    if (fields[name] == null) {
+                        fields[name] = FieldInfo(desc, hide, descriptor.type.isNullable(), null)
+                    } else {
+                        fields[name] = fields[name]!!.copy(desc = desc, needToHide = hide, isNullable = descriptor.type.isNullable())
+                    }
                 }
             }
         }
@@ -66,7 +72,6 @@ class SekretClassBuilder(
                     descriptor: String
                 ) {
                     if (opcode == Opcodes.GETFIELD) {
-                        println("skipping $owner/$name/$descriptor - opcode $opcode=='GET FIELD'")
                         fields[name] = null
                     }
                 }
@@ -83,8 +88,10 @@ class SekretClassBuilder(
                         && name.startsWith("get")
                     ) {
                         val keyName = name.substring(3).substringBefore("-").replaceFirstChar { it.lowercase() }
-                        println("skipping $owner/$keyName/$descriptor - opcode $opcode=='INVOKE VIRTUAL'. interface:$isInterface")
                         fields[keyName] = null
+                    } else if (opcode == Opcodes.INVOKESTATIC && name == TO_STRING_IMPL) {
+                        // lastly added field should be wrapped
+                        fields[fields.keys.last()] = FieldInfo("", false, false, owner)
                     }
                 }
             }
@@ -95,6 +102,9 @@ class SekretClassBuilder(
 
     override fun done() {
         if (generateToString) {
+            if (error.isNotEmpty()) {
+                throw java.lang.RuntimeException(error)
+            }
             generateToString()
         }
         super.done()
@@ -143,6 +153,16 @@ class SekretClassBuilder(
                 mv.visitFieldInsn(Opcodes.GETFIELD, thisName, name, fieldInfo.desc)
 
                 when {
+                    fieldInfo.wrapperClassName != null -> {
+                        mv.visitMethodInsn(
+                            Opcodes.INVOKESTATIC,
+                            fieldInfo.wrapperClassName,
+                            TO_STRING_IMPL,
+                            "(${fieldInfo.desc})Ljava/lang/String;",
+                            false,
+                        )
+                        appendToStringBuilder(mv, "Ljava/lang/Object;")
+                    }
                     // primitives or descriptors supported by StringBuilder
                     fieldInfo.desc.length == 1
                         || STRING_BUILDER_SUPPORTED_DESCRIPTORS.contains(fieldInfo.desc) -> {
@@ -198,6 +218,7 @@ class SekretClassBuilder(
 
     private companion object {
         const val STRING_BUILDER = "java/lang/StringBuilder"
+        const val TO_STRING_IMPL = "toString-impl"
         val STRING_BUILDER_SUPPORTED_DESCRIPTORS =
             setOf("Ljava/lang/String;", "Ljava/lang/StringBuffer;", "Ljava/lang/CharSequence;")
     }

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
@@ -26,8 +26,6 @@ class SekretClassBuilder(
     private val fields = linkedMapOf<String, FieldInfo?>()
     private var generateToString = false
 
-    private var error = ""
-
     override fun getDelegate(): ClassBuilder = classBuilder
 
     override fun newField(
@@ -107,9 +105,6 @@ class SekretClassBuilder(
 
     override fun done() {
         if (generateToString) {
-            if (error.isNotEmpty()) {
-                throw java.lang.RuntimeException(error)
-            }
             generateToString()
         }
         super.done()

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
@@ -82,7 +82,7 @@ class SekretClassBuilder(
                         && owner == origin.classDescriptor()
                         && name.startsWith("get")
                     ) {
-                        val keyName = name.substring(3).replaceFirstChar { it.lowercase() }
+                        val keyName = name.substring(3).substringBefore("-").replaceFirstChar { it.lowercase() }
                         println("skipping $owner/$keyName/$descriptor - opcode $opcode=='INVOKE VIRTUAL'. interface:$isInterface")
                         fields[keyName] = null
                     }
@@ -112,11 +112,7 @@ class SekretClassBuilder(
         appendToStringBuilder(mv)
 
         fields.onEachIndexed { index, (name, fieldInfo) ->
-
-            if (fieldInfo == null) {
-                println("WARNING $index/$name has no field info")
-                return@onEachIndexed
-            }
+            checkNotNull(fieldInfo) { name }
 
             mv.visitLdcInsn((if (index == 0) "" else ", ") + "$name=")
             appendToStringBuilder(mv)
@@ -153,7 +149,7 @@ class SekretClassBuilder(
                         appendToStringBuilder(mv, fieldInfo.desc)
                     }
                     // arrays
-                    fieldInfo.desc[0] == '['                                             -> {
+                    fieldInfo.desc[0] == '[' -> {
                         val isPrimitiveArray = fieldInfo.desc.length == 2
                         mv.visitMethodInsn(
                             Opcodes.INVOKESTATIC,
@@ -165,7 +161,7 @@ class SekretClassBuilder(
                         appendToStringBuilder(mv)
                     }
                     // others go as Object
-                    else                                                                 -> {
+                    else -> {
                         appendToStringBuilder(mv, "Ljava/lang/Object;")
                     }
                 }

--- a/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
+++ b/kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt
@@ -91,7 +91,12 @@ class SekretClassBuilder(
                         fields[keyName] = null
                     } else if (opcode == Opcodes.INVOKESTATIC && name == TO_STRING_IMPL) {
                         // lastly added field should be wrapped
-                        fields[fields.keys.last()] = FieldInfo("", false, false, owner)
+                        fields[fields.keys.last()] = FieldInfo(
+                            desc = "",
+                            needToHide = false,
+                            isNullable = false,
+                            wrapperClassName = owner
+                        )
                     }
                 }
             }

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/App.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/App.kt
@@ -91,4 +91,8 @@ fun main() {
             student = Student("student", "pwd"),
         )
     )
+
+    Interface.main()
+    VC.main()
+
 }

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/interfaces.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/interfaces.kt
@@ -1,0 +1,52 @@
+package net.afanasev.sekret.sample
+
+import net.afanasev.sekret.Secret
+
+@Suppress("ArrayInDataClass")
+data class ComplexImpl(
+    @Secret override val str: String,
+    @Secret override val int: Int,
+    @Secret override val long: Long,
+    @Secret override val float: Float,
+    @Secret override val list: List<String>,
+    @Secret override val array: Array<Char>,
+    @Secret override var boolean: Boolean,
+    override val admin: Admin,
+    @Secret override val student: Student,
+) : ComplexInterface {
+    companion object {
+        const val CONST = "const"
+    }
+
+    val field1 = ""
+    var field2 = false
+}
+
+interface ComplexInterface {
+
+    val str: String
+    val int: Int
+    val long: Long
+    val float: Float
+    val list: List<String>
+    val array: Array<Char>
+    var boolean: Boolean
+    val admin: Admin
+    val student: Student
+}
+
+fun main() {
+    println(
+        ComplexImpl(
+            str = "string",
+            int = 3,
+            long = 4L,
+            float = 2f,
+            list = listOf(),
+            array = arrayOf(),
+            boolean = true,
+            admin = Admin("admin", "pwd"),
+            student = Student("student", "pwd"),
+        )
+    )
+}

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/interfaces.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/interfaces.kt
@@ -35,18 +35,20 @@ interface ComplexInterface {
     val student: Student
 }
 
-fun main() {
-    println(
-        ComplexImpl(
-            str = "string",
-            int = 3,
-            long = 4L,
-            float = 2f,
-            list = listOf(),
-            array = arrayOf(),
-            boolean = true,
-            admin = Admin("admin", "pwd"),
-            student = Student("student", "pwd"),
+object Interface {
+    fun main() {
+        println(
+            ComplexImpl(
+                str = "string",
+                int = 3,
+                long = 4L,
+                float = 2f,
+                list = listOf(),
+                array = arrayOf(),
+                boolean = true,
+                admin = Admin("admin", "pwd"),
+                student = Student("student", "pwd"),
+            )
         )
-    )
+    }
 }

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
@@ -12,12 +12,16 @@ value class Name(val value: String) {
 }
 
 @JvmInline
+value class NestedName(val name: Name)
+
+@JvmInline
 value class Description(val value: String)
 
 // regular data class - expect Secket has no effect
 data class MyModel(
     val id: ModelId,
     val name: Name,
+    val nestedName: NestedName,
     val description: Description?,
 )
 
@@ -25,6 +29,7 @@ data class MyModel(
 data class MyModelSecret(
     @Secret val id: ModelId,
     @Secret val name: Name,
+    @Secret val nestedName: NestedName,
     @Secret val description: Description?,
 )
 
@@ -32,6 +37,7 @@ data class MyModelSecret(
 data class MySubModel(
     override val id: ModelId,
     override val name: Name,
+    override val nestedName: NestedName,
     override val description: Description?,
     val extraField: String,
 ) : MyModelInterface
@@ -40,6 +46,7 @@ data class MySubModel(
 data class MySubModelSecret(
     @Secret override val id: ModelId,
     @Secret override val name: Name,
+    @Secret override val nestedName: NestedName,
     @Secret override val description: Description?,
     @Secret val anotherExtraField: String,
 ) : MyModelInterface
@@ -47,29 +54,27 @@ data class MySubModelSecret(
 interface MyModelInterface {
     val id: ModelId
     val name: Name
+    val nestedName: NestedName
     val description: Description?
 }
 
-fun main() {
-
-}
 object VC {
     fun main() {
         val modelId = ModelId(111)
         val name = Name("name")
+        val nestedName = NestedName(Name("nested"))
         val description = Description("description")
 
         println(modelId)
         println(name)
         println(description)
 
-        println(MyModel(modelId, name, description))
+        println(MyModel(modelId, name, nestedName, description))
 
-        println(MyModelSecret(modelId, name, description))
+        println(MyModelSecret(modelId, name, nestedName, description))
 
-        println(MySubModel(modelId, name, description, "extra-field"))
+        println(MySubModel(modelId, name, nestedName, description, "extra-field"))
 
-        println(MySubModelSecret(modelId, name, description, "another-extra-field"))
-
+        println(MySubModelSecret(modelId, name, nestedName, description, "another-extra-field"))
     }
 }

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
@@ -5,8 +5,10 @@ import net.afanasev.sekret.Secret
 
 @JvmInline
 value class ModelId(val value: Long) : Comparable<Long> by value
+
 @JvmInline
 value class Name(val value: String) : CharSequence by value
+
 @JvmInline
 value class Description(val value: String)
 
@@ -47,20 +49,25 @@ interface MyModelInterface {
 }
 
 fun main() {
-    val modelId = ModelId(111)
-    val name = Name("name")
-    val description = Description("description")
 
-    println(modelId)
-    println(name)
-    println(description)
+}
+object VC {
+    fun main() {
+        val modelId = ModelId(111)
+        val name = Name("name")
+        val description = Description("description")
 
-    println(MyModel(modelId, name, description))
+        println(modelId)
+        println(name)
+        println(description)
 
-    println(MyModelSecret(modelId, name, description))
+        println(MyModel(modelId, name, description))
 
-    println(MySubModel(modelId, name, description, "extra-field"))
+        println(MyModelSecret(modelId, name, description))
 
-    println(MySubModelSecret(modelId, name, description, "another-extra-field"))
+        println(MySubModel(modelId, name, description, "extra-field"))
 
+        println(MySubModelSecret(modelId, name, description, "another-extra-field"))
+
+    }
 }

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
@@ -7,7 +7,9 @@ import net.afanasev.sekret.Secret
 value class ModelId(val value: Long) : Comparable<Long> by value
 
 @JvmInline
-value class Name(val value: String) : CharSequence by value
+value class Name(val value: String) {
+    override fun toString(): String = "my name is $value"
+}
 
 @JvmInline
 value class Description(val value: String)

--- a/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
+++ b/sample/src/main/kotlin/net/afanasev/sekret/sample/valueClasses.kt
@@ -1,0 +1,66 @@
+package net.afanasev.sekret.sample
+
+import net.afanasev.sekret.Secret
+
+
+@JvmInline
+value class ModelId(val value: Long) : Comparable<Long> by value
+@JvmInline
+value class Name(val value: String) : CharSequence by value
+@JvmInline
+value class Description(val value: String)
+
+// regular data class - expect Secket has no effect
+data class MyModel(
+    val id: ModelId,
+    val name: Name,
+    val description: Description?,
+)
+
+// Secket should censor all @Secret fields from toString()
+data class MyModelSecret(
+    @Secret val id: ModelId,
+    @Secret val name: Name,
+    @Secret val description: Description?,
+)
+
+// regular data class - expect Secket has no effect
+data class MySubModel(
+    override val id: ModelId,
+    override val name: Name,
+    override val description: Description?,
+    val extraField: String,
+) : MyModelInterface
+
+// Secket should censor all @Secret fields from toString()
+data class MySubModelSecret(
+    @Secret override val id: ModelId,
+    @Secret override val name: Name,
+    @Secret override val description: Description?,
+    @Secret val anotherExtraField: String,
+) : MyModelInterface
+
+interface MyModelInterface {
+    val id: ModelId
+    val name: Name
+    val description: Description?
+}
+
+fun main() {
+    val modelId = ModelId(111)
+    val name = Name("name")
+    val description = Description("description")
+
+    println(modelId)
+    println(name)
+    println(description)
+
+    println(MyModel(modelId, name, description))
+
+    println(MyModelSecret(modelId, name, description))
+
+    println(MySubModel(modelId, name, description, "extra-field"))
+
+    println(MySubModelSecret(modelId, name, description, "another-extra-field"))
+
+}

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/InterfaceSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/InterfaceSpec.kt
@@ -12,15 +12,15 @@ object InterfaceSpec : Spek({
         it("annotated properties should be hidden") {
             assertEquals(
                 "ComplexImpl(" +
-                    "str=■■■, " +
-                    "int=■■■, " +
-                    "long=■■■, " +
-                    "float=■■■, " +
-                    "list=■■■, " +
-                    "array=■■■, " +
-                    "boolean=■■■, " +
-                    "admin=Admin(login=admin, password=■■■), " +
-                    "student=■■■" +
+                    "str=$mask, " +
+                    "int=$mask, " +
+                    "long=$mask, " +
+                    "float=$mask, " +
+                    "list=$mask, " +
+                    "array=$mask, " +
+                    "boolean=$mask, " +
+                    "admin=Admin(login=admin, password=$mask), " +
+                    "student=$mask" +
                     ")",
                 ComplexImpl(
                     str = "string",

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/InterfaceSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/InterfaceSpec.kt
@@ -1,0 +1,39 @@
+package net.afanasev.sekret.sample
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+object InterfaceSpec : Spek({
+
+    describe("data class that implements an interface") {
+        val mask = "■■■"
+
+        it("annotated properties should be hidden") {
+            assertEquals(
+                "ComplexImpl(" +
+                    "str=■■■, " +
+                    "int=■■■, " +
+                    "long=■■■, " +
+                    "float=■■■, " +
+                    "list=■■■, " +
+                    "array=■■■, " +
+                    "boolean=■■■, " +
+                    "admin=Admin(login=admin, password=■■■), " +
+                    "student=■■■" +
+                    ")",
+                ComplexImpl(
+                    str = "string",
+                    int = 3,
+                    long = 4L,
+                    float = 2f,
+                    list = listOf(),
+                    array = arrayOf(),
+                    boolean = true,
+                    admin = Admin("admin", "pwd"),
+                    student = Student("student", "pwd"),
+                ).toString()
+            )
+        }
+    }
+})

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/SekretSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/SekretSpec.kt
@@ -33,6 +33,37 @@ object SekretSpec : Spek({
         it("should hide getters") {
             assertEquals("BaseImpl(str=$mask, id=$mask, arr=$mask, str2=$mask, str3=str3)", BaseImpl("str", 42, arrayOf("a", "r"), "str3", "str3").toString())
         }
+
+        it("should hide getters") {
+            assertEquals("BaseImpl(str=$mask, id=$mask, arr=$mask, str2=$mask, str3=str3)", BaseImpl("str", 42, arrayOf("a", "r"), "str3", "str3").toString())
+        }
+
+        it("should hide properties of a complex object") {
+            assertEquals(
+                "Complex(" +
+                    "str=■■■, " +
+                    "int=■■■, " +
+                    "long=■■■, " +
+                    "float=■■■, " +
+                    "list=■■■, " +
+                    "array=■■■, " +
+                    "boolean=■■■, " +
+                    "admin=Admin(login=admin, password=■■■), " +
+                    "student=■■■" +
+                    ")",
+                Complex(
+                    str = "string",
+                    int = 3,
+                    long = 4L,
+                    float = 2f,
+                    list = listOf(),
+                    array = arrayOf(),
+                    boolean = true,
+                    admin = Admin("admin", "pwd"),
+                    student = Student("student", "pwd"),
+                ).toString()
+            )
+        }
     }
 
 })

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/SekretSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/SekretSpec.kt
@@ -41,15 +41,15 @@ object SekretSpec : Spek({
         it("should hide properties of a complex object") {
             assertEquals(
                 "Complex(" +
-                    "str=■■■, " +
-                    "int=■■■, " +
-                    "long=■■■, " +
-                    "float=■■■, " +
-                    "list=■■■, " +
-                    "array=■■■, " +
-                    "boolean=■■■, " +
-                    "admin=Admin(login=admin, password=■■■), " +
-                    "student=■■■" +
+                    "str=$mask, " +
+                    "int=$mask, " +
+                    "long=$mask, " +
+                    "float=$mask, " +
+                    "list=$mask, " +
+                    "array=$mask, " +
+                    "boolean=$mask, " +
+                    "admin=Admin(login=admin, password=$mask), " +
+                    "student=$mask" +
                     ")",
                 Complex(
                     str = "string",

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
@@ -7,6 +7,7 @@ import org.spekframework.spek2.style.specification.describe
 object ValueClassSpec : Spek({
     val modelId = ModelId(111)
     val name = Name("name")
+    val nestedName = NestedName(Name("nested"))
     val description = Description("description")
 
     describe("data class with value class properties") {
@@ -14,15 +15,25 @@ object ValueClassSpec : Spek({
 
         it("should print non-annotated properties") {
             assertEquals(
-                "MyModel(id=ModelId(value=111), name=my name is name, description=Description(value=description))",
-                MyModel(modelId, name, description).toString()
+                "MyModel(" +
+                    "id=ModelId(value=111), " +
+                    "name=my name is name, " +
+                    "nestedName=NestedName(name=my name is nested), " +
+                    "description=Description(value=description)" +
+                    ")",
+                MyModel(modelId, name, nestedName, description).toString()
             )
         }
 
         it("should hide annotated properties") {
             assertEquals(
-                "MyModelSecret(id=$mask, name=$mask, description=$mask)",
-                MyModelSecret(modelId, name, description).toString()
+                "MyModelSecret(" +
+                    "id=$mask, " +
+                    "name=$mask, " +
+                    "nestedName=$mask, " +
+                    "description=$mask" +
+                    ")",
+                MyModelSecret(modelId, name, nestedName, description).toString()
             )
         }
 
@@ -30,15 +41,27 @@ object ValueClassSpec : Spek({
 
             it("should print non-annotated properties") {
                 assertEquals(
-                    "MySubModel(id=ModelId(value=111), name=my name is name, description=Description(value=description), extraField=extra-field)",
-                    MySubModel(modelId, name, description, "extra-field").toString()
+                    "MySubModel(" +
+                        "id=ModelId(value=111), " +
+                        "name=my name is name, " +
+                        "nestedName=NestedName(name=my name is nested), " +
+                        "description=Description(value=description), " +
+                        "extraField=extra-field" +
+                        ")",
+                    MySubModel(modelId, name, nestedName, description, "extra-field").toString()
                 )
             }
 
             it("should hide annotated properties") {
                 assertEquals(
-                    "MySubModelSecret(id=$mask, name=$mask, description=$mask, anotherExtraField=$mask)",
-                    MySubModelSecret(modelId, name, description, "another-extra-field").toString()
+                    "MySubModelSecret(" +
+                        "id=$mask, " +
+                        "name=$mask, " +
+                        "nestedName=$mask, " +
+                        "description=$mask, " +
+                        "anotherExtraField=$mask" +
+                        ")",
+                    MySubModelSecret(modelId, name, nestedName, description, "another-extra-field").toString()
                 )
             }
         }

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
@@ -1,0 +1,47 @@
+package net.afanasev.sekret.sample
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+object ValueClassSpec : Spek({
+    val modelId = ModelId(111)
+    val name = Name("name")
+    val description = Description("description")
+
+    describe("data class with value class properties") {
+        val mask = "■■■"
+
+        it("should print properties that are value classes, and not annotated") {
+            assertEquals(
+                "MyModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description))",
+                MyModel(modelId, name, description).toString()
+            )
+        }
+
+        it("should hide properties that are value classes, and are annotated") {
+            assertEquals(
+                "MyModelSecret(id=$mask, name=$mask, description=$mask)",
+                MyModelSecret(modelId, name, description).toString()
+            )
+        }
+
+        describe("that implement an interface") {
+
+            it("should print properties that are value classes, and not annotated") {
+                assertEquals(
+                    "MySubModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description), extraField=extra-field)",
+                    MySubModel(modelId, name, description, "extra-field").toString()
+                )
+            }
+
+            it("should hide properties that are value classes, and are annotated") {
+                assertEquals(
+                    "MySubModelSecret(id=$mask, name=$mask, description=$mask, anotherExtraField=$mask)",
+                    MySubModelSecret(modelId, name, description, "another-extra-field").toString()
+                )
+            }
+        }
+    }
+
+})

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
@@ -12,14 +12,14 @@ object ValueClassSpec : Spek({
     describe("data class with value class properties") {
         val mask = "■■■"
 
-        it("should print properties that are value classes, and not annotated") {
+        it("should print non-annotated properties") {
             assertEquals(
                 "MyModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description))",
                 MyModel(modelId, name, description).toString()
             )
         }
 
-        it("should hide properties that are value classes, and are annotated") {
+        it("should hide annotated properties") {
             assertEquals(
                 "MyModelSecret(id=$mask, name=$mask, description=$mask)",
                 MyModelSecret(modelId, name, description).toString()
@@ -28,14 +28,14 @@ object ValueClassSpec : Spek({
 
         describe("that implement an interface") {
 
-            it("should print properties that are value classes, and not annotated") {
+            it("should print non-annotated properties") {
                 assertEquals(
                     "MySubModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description), extraField=extra-field)",
                     MySubModel(modelId, name, description, "extra-field").toString()
                 )
             }
 
-            it("should hide properties that are value classes, and are annotated") {
+            it("should hide annotated properties") {
                 assertEquals(
                     "MySubModelSecret(id=$mask, name=$mask, description=$mask, anotherExtraField=$mask)",
                     MySubModelSecret(modelId, name, description, "another-extra-field").toString()

--- a/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
+++ b/sample/src/test/kotlin/net/afanasev/sekret/sample/ValueClassSpec.kt
@@ -14,7 +14,7 @@ object ValueClassSpec : Spek({
 
         it("should print non-annotated properties") {
             assertEquals(
-                "MyModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description))",
+                "MyModel(id=ModelId(value=111), name=my name is name, description=Description(value=description))",
                 MyModel(modelId, name, description).toString()
             )
         }
@@ -30,7 +30,7 @@ object ValueClassSpec : Spek({
 
             it("should print non-annotated properties") {
                 assertEquals(
-                    "MySubModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description), extraField=extra-field)",
+                    "MySubModel(id=ModelId(value=111), name=my name is name, description=Description(value=description), extraField=extra-field)",
                     MySubModel(modelId, name, description, "extra-field").toString()
                 )
             }

